presto-main/src/main/java/io/prestosql/operator/FilterAndProjectOperator.java

```diff
@@ -15,95 +15,89 @@
+import io.prestosql.operator.WorkProcessorOperatorAdapter.AdapterWorkProcessorOperatorFactory;
 import io.prestosql.operator.project.PageProcessor;
 import io.prestosql.spi.Page;
 import io.prestosql.spi.type.Type;
 import io.prestosql.sql.planner.plan.PlanNodeId;
 
 import java.util.List;
+import java.util.Optional;
 import java.util.function.Supplier;
 
 import static com.google.common.base.Preconditions.checkState;
 import static io.prestosql.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static io.prestosql.operator.project.MergePages.mergePages;
 import static java.util.Objects.requireNonNull;
 
 public class FilterAndProjectOperator
-        implements Operator
+        implements AdapterWorkProcessorOperator
 {
-    private final OperatorContext operatorContext;
-    private final LocalMemoryContext pageProcessorMemoryContext;
-    private final LocalMemoryContext outputMemoryContext;
-
-    private final PageProcessor processor;
-    private final MergingPageOutput mergingOutput;
-    private boolean finishing;
+    private final PageBuffer pageBuffer = new PageBuffer();
+    private final WorkProcessor<Page> pages;
 
     public FilterAndProjectOperator(
-            OperatorContext operatorContext,
-            PageProcessor processor,
-            MergingPageOutput mergingOutput)
-    {
-        this.processor = requireNonNull(processor, "processor is null");
-        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
-        this.pageProcessorMemoryContext = newSimpleAggregatedMemoryContext().newLocalMemoryContext(ScanFilterAndProjectOperator.class.getSimpleName());
-        this.outputMemoryContext = operatorContext.newLocalSystemMemoryContext(FilterAndProjectOperator.class.getSimpleName());
-        this.mergingOutput = requireNonNull(mergingOutput, "mergingOutput is null");
-    }
-
-    @Override
-    public OperatorContext getOperatorContext()
+            Session session,
+            MemoryTrackingContext memoryTrackingContext,
+            DriverYieldSignal yieldSignal,
+            Optional<WorkProcessor<Page>> sourcePages,
+            PageProcessor pageProcessor,
+            List<Type> types,
+            DataSize minOutputPageSize,
+            int minOutputPageRowCount)
     {
-        return operatorContext;
+        AggregatedMemoryContext aggregatedMemoryContext = memoryTrackingContext.aggregateUserMemoryContext();
+        LocalMemoryContext memoryContext = aggregatedMemoryContext.newLocalMemoryContext(FilterAndProjectOperator.class.getSimpleName());
+        AggregatedMemoryContext localAggregatedMemoryContext = newSimpleAggregatedMemoryContext();
+        LocalMemoryContext outputMemoryContext = localAggregatedMemoryContext.newLocalMemoryContext(FilterAndProjectOperator.class.getSimpleName());
+
+        this.pages = sourcePages.orElse(pageBuffer.pages())
+                .yielding(yieldSignal::isSet)
```

Comment:
this is unneeded. it will be added automatically by `WorkProcessorPipelineSourceOperator` between operators. Also yielding before `pageBuffer` doesn't make sense as fetching data from it is inexpensive